src/armature/CostFunction.ts

```diff
@@ -0,0 +1,182 @@
+import { coord } from '../calder';
+import { AABB } from '../geometry/BakedGeometry';
+import { vec3From4 } from '../math/utils';
+import { Mapper } from '../utils/mapper';
+import { CostFn, GeneratorInstance } from './Generator';
+import { Model } from './Model';
+import { GeometryNode, Node } from './Node';
+
+import { vec3, vec4 } from 'gl-matrix';
+import { range } from 'lodash';
+
+export type ForcePoint = {
+    point: coord;
+    influence: number;
+};
+
+type Grid = { [key: string]: true };
+
+export namespace CostFunction {
+    /**
+     * Use points with positive and negative influence to control generation.
+     *
+     * @param {ForcePoint[]} points The points of influence, where negative influence means the
+     * point reduces the overall cost when nodes get close.
+     * @returns {CostFn} The resulting cost function.
+     */
+    export function forces(points: ForcePoint[]): CostFn {
+        const vectors = points.map((forcePoint: ForcePoint) => {
+            return {
+                vector: Mapper.coordToVector(forcePoint.point),
+                influence: forcePoint.influence
+            };
+        });
+
+        return (instance: GeneratorInstance, added: Node[]) => {
+            // Out of the added nodes, just get the geometry nodes
+            const addedGeometry: GeometryNode[] = [];
+            added.forEach((n: Node) =>
+                n.geometryCallback((node: GeometryNode) => {
+                    addedGeometry.push(node);
+                })
+            );
+
+            // For each added shape and each influence point, add the resulting cost to the
+            // instance's existing cost.
+            return addedGeometry.reduce((sum: number, node: GeometryNode) => {
```

Comment:
Nested `reduce`s can be hard to comprehend upon initial glance, but I am a fan of doing things functionally. I'm not sure what you think is a better tradeoff - doing this functionally and having less readability or doing imperatively with more readability.

Reply:
Probably better to just go with a forEach that mutates a variable, it'll probably be clearer